orttraining/orttraining/python/training/ort_triton/_common.py

```diff
@@ -0,0 +1,188 @@
+    return [input_infos[0], TensorInfo(TensorProto.BOOL, input_infos[0].shape)]
+
+
+class TypeAndShapeInfer:
+    _INFER_FUNC_MAP = {
+        "Add": _infer_elementwise,
+        "Sub": _infer_elementwise,
+        "Mul": _infer_elementwise,
+        "Div": _infer_elementwise,
+        "Pow": _infer_elementwise,
+        "Sqrt": _infer_elementwise,
+        "Exp": _infer_elementwise,
+        "Where": _infer_where,
+        "Rsqrt": _infer_elementwise,
+        "Cast": _infer_cast,
+        "Dropout": _infer_dropout,
+        "DropoutGrad": _infer_unary,
+        "Identity": _infer_unary,
+        "ReduceSum": _infer_reduction,
+        "ReduceMax": _infer_reduction,
+        "ReduceMin": _infer_reduction,
+        "Sum": _infer_elementwise,
+    }
+
+    @classmethod
+    def infer(cls, node: NodeProto, input_infos: List[TensorInfo], graph: GraphProto) -> List[TensorInfo]:
+        if node.op_type not in cls._INFER_FUNC_MAP:
+            raise NotImplementedError(f"Unsupported op type: {node.op_type}")
+        return cls._INFER_FUNC_MAP[node.op_type](node, input_infos, graph)
+
+
+class AutotuneConfigs:
+    """
+    Generate all autotune configs for a kernel function by it's xnumel and rnumel.
+    A config is a tuple of (xblock, rblock, num_warps).
+    If it's elementwise kernel, the rnumel is 1.
+    If it's reduction kernel on last contiguous dimensions, the contiguous flag is True.
+    """
+
+    def __init__(self, x_numel: int, r_numel: int, contiguous: bool):
+        self.configs: List[Tuple[int, int, int]] = self._gen_autotune_configs(x_numel, r_numel, contiguous)
+        self.requires_for_loop: bool = any(config[1] < r_numel for config in self.configs)
+
+    def _num_warps(self, x: int, r: int) -> int:
+        return min(max(x * r // 256, 2), 8)
+
+    def _gen_config(self, xnp2: int, rnp2: int, x: int, r: int) -> Tuple[int, int, int]:
+        x = min(x, xnp2)
+        r = min(r, rnp2)
+        return x, r, self._num_warps(x, r)
+
+    # TODO: we need to tune more kernels to get more reasonable configs for better performance.
+    def _gen_autotune_configs(self, x_numel: int, r_numel: int, contiguous: bool) -> List[Tuple[int, int, int]]:
+        configs = []
+        xnp2 = next_power_of_2(x_numel)
+        if r_numel == 1:
+            configs.append(self._gen_config(xnp2, 1, 1024, 1))
+            if xnp2 > 1024:
+                configs.append(self._gen_config(xnp2, 1, 2048, 1))
+            return configs
+        rnp2 = next_power_of_2(r_numel)
+        if contiguous:
+            configs.append(self._gen_config(xnp2, rnp2, 1, 2048))
+            if rnp2 > 2048:
+                configs.append(self._gen_config(xnp2, rnp2, 1, 4096))
+            elif rnp2 <= 256:
+                x = min(xnp2, 256 // rnp2 * 2)
+                configs.append(self._gen_config(xnp2, rnp2, x, rnp2))
+        else:
+            config_set = {
+                self._gen_config(xnp2, rnp2, 1, 2048),
```

Comment:
Do we expect those figures to change in the future with new machines?

Reply:
Currently this config is mainly  following some logic from PyTorch, and currently it's tested on V100 only. It's possible we will adjust this piece of code in the future by doing more testing and tuning on different devices. Adding more tuning config can help to get better result, but it will slow down the tuning perf a lot, need more time to get the balance in the future.

Reply:
Maybe then adding some comments to tell where these values are coming from.